src/main/java/com/hotels/hcommon/hive/metastore/paths/PartitionedTablePathResolver.java

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (C) 2018-2019 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.hcommon.hive.metastore.paths;
+
+import static com.hotels.hcommon.hive.metastore.util.LocationUtils.locationAsPath;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.Warehouse;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hotels.hcommon.hive.metastore.iterator.PartitionIterator;
+
+class PartitionedTablePathResolver implements TablePathResolver {
+
+  private static final Logger log = LoggerFactory.getLogger(PartitionedTablePathResolver.class);
+
+  private final Path tableBaseLocation;
+  private final Path globPath;
+  private final IMetaStoreClient metaStore;
+  private final Table table;
+
+  PartitionedTablePathResolver(IMetaStoreClient metaStore, Table table)
+      throws NoSuchObjectException, MetaException, TException {
+    this.metaStore = metaStore;
+    this.table = table;
+    log.debug("Table '{}' is partitioned", Warehouse.getQualifiedName(table));
+    tableBaseLocation = locationAsPath(table);
+    List<Partition> onePartition = metaStore.listPartitions(table.getDbName(), table.getTableName(), (short) 1);
+    if (onePartition.isEmpty()) {
+      String errorMessage = String.format("Table '%s' has no partitions, perhaps you can simply delete: %s.",
```

Comment:
Since this is now being used outside of housekeeping it probably doesn't make sense to talk about deletion. The error message should just indicate that it has no partitions (i.e. WTF are you doing trying to use this partitioned resolver on a table with no partitions).

Reply:
Agree, although it would be the factory deciding what class to use, so if we get to that error message, the table has partition keys, but they're empty. Can that happen?

Reply:
I guess the `getPartitionKeys()` (i.e. "yes this table is partitioned") could return something but `listPartitions` could return empty (i.e. "but none have been added"). Edge case, not sure how useful it is but let's just keep this for now as its backwards compatible and we change it in future if we decide we don't need this behaviour here (in which case we'd probably need to move the check into the vacuum tool if it was the only place that actually needed it).